### PR TITLE
Rework publish scripts

### PR DESCRIPTION
**Pull request by frant-hartm**
The goal is to be able to publish packages for both OS and EE and for any versions ve release (stable, snapshot, development releases and betas).

I have moved all the code to prepare the package to `build-[deb|rpm|homebrew].sh` scripts. These are then called from GH action `publish-packages.yaml`

The scrips use the following env variables as input:

HZ_DISTRIBUTION - `hazelcast` or `hazelcast-enterprise` for OS or EE version of the distribution package
HZ_VERSION - version of hazelcast
PACKAGE_VERSION - version of the package based on HZ_VERSION
PUBLISH - if to upload final package to the repository

HZ-536
HZ-533
HZ-508

TODOs:

- [x] move HZ_VERSION to PACKAGE_VERSION conversion to common.sh
- [ ] update readme
- [ ] add some documentation (started here https://docs.google.com/document/d/1fIC7hzZDMXg7PN_Kvyxg1r3kQuBqvLjG-kVNJPy1uyQ/edit?usp=sharing)
- [ ] change to production repositories when all is done and testing finished
- [ ] upgrade on homebrew alias doesn't seem to work
- [ ] homebrew `conflicts_with` doesn't work as I expect
- [ ] should we make this repository private?

### CODE REVIEW

File: build-homebrew-package.sh [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  export MAVEN_REPO="https://repository.hazelcast.com/devel"
+  export PACKAGE_URL="$MAVEN_REPO/com/hazelcast/${HZ_DISTRIBUTION}-distribution/${HZ_VERSION}/${HZ_DISTRIBUTION}-distribution-${HZ_VERSION}.tar.gz"
+else
+  if [ "${HZ_DISTRIBUTION}" == "hazelcast" ]; then
+    export MAVEN_REPO="https://repo1.maven.org/maven2"
+  else
+    export MAVEN_REPO="https://repository.hazelcast.com/release"
+  fi
+  export PACKAGE_URL="$MAVEN_REPO/com/hazelcast/${HZ_DISTRIBUTION}-distribution/${HZ_VERSION}/${HZ_DISTRIBUTION}-distribution-${HZ_VERSION}.tar.gz"
+fi
```
Comment (ldziedziul) [inline build-homebrew-package.sh:45]: This is quite complicated. I wonder if we could ask maven for the package URL. A bit hacky way would be something like:
```bash
mvn dependency:get -Dartifact=-Dartifact=com.hazelcast:${HZ_DISTRIBUTION}-distribution:${HZ_VERSION}:tar.gz -Dtransitive=false | grep 'Downloaded from' | grep -Eo 'https://[^ >]+'
```
What do you think?

File: build-homebrew-package.sh [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+ASSET_SHASUM=$(sha256sum ${HZ_DISTRIBUTION}-distribution-${HZ_VERSION}.tar.gz | cut -d ' ' -f 1)
+
+# If this is 'hazelcast' package it conflicts with 'hazelcast-enterprise' and vice versa
+export CONFLICTS=hazelcast-enterprise
+if [ ${HZ_DISTRIBUTION} == "hazelcast-enterprise" ]; then
+  export CONFLICTS=hazelcast
+fi
+
+if [ -d homebrew-hz ]; then
```
Comment (ldziedziul) [inline build-homebrew-package.sh:59]: Why this `if`?

File: packages/deb/hazelcast/DEBIAN/conffiles [hunk 3]
```diff
@@ -1,12 +1,14 @@
...
-/usr/lib/hazelcast/hazelcast-4.1-BETA-1/config/examples/hazelcast-full-example.yaml
-/usr/lib/hazelcast/hazelcast-4.1-BETA-1/config/examples/hazelcast.yaml
-/usr/lib/hazelcast/hazelcast-4.1-BETA-1/config/examples/hazelcast-client.yaml
-/usr/lib/hazelcast/hazelcast-4.1-BETA-1/config/hazelcast.xml
-/usr/lib/hazelcast/hazelcast-4.1-BETA-1/config/hazelcast-client.xml
-/usr/lib/hazelcast/hazelcast-4.1-BETA-1/config/jmx_agent_config.yaml
-/usr/lib/hazelcast/hazelcast-4.1-BETA-1/config/jvm-client.options
-/usr/lib/hazelcast/hazelcast-4.1-BETA-1/config/jvm.options
-/usr/lib/hazelcast/hazelcast-4.1-BETA-1/config/log4j2.properties
+/usr/lib/hazelcast/${HZ_DISTRIBUTION}-${HZ_VERSION}/config/examples/hazelcast-client-full-example.xml
```
Comment (ldziedziul) [inline packages/deb/hazelcast/DEBIAN/conffiles:1]: ❤️

File: packages/rpm/hazelcast.spec [hunk 4]
```diff
@@ -63,7 +65,15 @@ if [  ! -f %{buildroot}/%{_bindir}/hz  ]; then
 fi
 
 %files
-%{_prefix}/lib/%{name}/%{name}-%{hzversion}/*
+# The LICENSE file contains Apache 2 license ans is only present in OS
```
Comment (ldziedziul) [inline packages/rpm/hazelcast.spec:68]: typ in `ans`?

File: .github/workflows/publish-packages.yml [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+      - uses: actions/cache@v2
+        with:
+          path: ~/.m2/repository
+          key: ${{ runner.os }}-maven-${{ hashFiles('**/pom.xml') }}
+          restore-keys: ${{ runner.os }}-maven-
+
+      - name: Set HZ_VERSION
+        run: |
+          if [ -z "${{ env.HZ_VERSION }}" ]; then
+            echo "HZ_VERSION=$(grep '<hazelcast.version>' pom.xml | sed "s@.*<hazelcast.version>\(.*\)</hazelcast.version>.*@\1@")" >> $GITHUB_ENV
```
[the remark this answers is not in the document]
Comment (frant-hartm) [reply .github/workflows/publish-packages.yml:49]: Love this.

File: .github/workflows/publish-packages.yml [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+    steps:
+      - uses: actions/checkout@v2.3.2
+
+      # TODO change debian-test-local to debian-local before merge when we finish testing
+      - name: Install CLI from deb
+        run: |
+          source ./common.sh
+          derive_package_repo "${HZ_VERSION}"
+          # await some time for deb metadata to be signed
+          sleep 120
```
[the remark this answers is not in the document]
Comment (frant-hartm) [reply .github/workflows/publish-packages.yml:126]: I actually don't know exactly how this works, or if the packages are signed at all. Kept what there was previously.

[the remark this answers is not in the document]
Comment (frant-hartm) [reply .github/workflows/publish-packages.yml:126]: I added call to REST API to recalculate metadata and removed this sleep.

File: .github/workflows/publish-packages.yml [hunk 7]
```diff
@@ -7,12 +7,188 @@ on:
...
+            PACKAGE_VERSION=$(echo ${{ github.ref }} | cut -c 12-)
+          else
+            PACKAGE_VERSION=${{ env.HZ_VERSION }}
+          fi
+          echo "PACKAGE_VERSION=$PACKAGE_VERSION" >> $GITHUB_ENV
+          echo ::set-output name=package_version::$PACKAGE_VERSION
+
+      - name: Download the distribution tar.gz file
+        run: |
+          HZ_PACKAGE_URL=$(mvn --batch-mode dependency:copy -Dartifact=com.hazelcast:${HZ_DISTRIBUTION}-distribution:${HZ_VERSION}:tar.gz -DoutputDirectory=./ -Dmdep.useBaseVersion=true | grep 'Downloaded from.*hazelcast.*' | grep -v 'maven-metadata.xml' | grep -Eo 'https://[^ >]+' | tail -1)
```
Comment (ldziedziul) [inline .github/workflows/publish-packages.yml:67]: What about this? 
```suggestion
          HZ_PACKAGE_URL=$(mvn --batch-mode dependency:copy -Dartifact=com.hazelcast:${HZ_DISTRIBUTION}-distribution:${HZ_VERSION}:tar.gz -DoutputDirectory=./ -Dmdep.useBaseVersion=true | grep 'Downloaded from' | grep -Eo 'https://[^ >]+${HZ_DISTRIBUTION}-distribution-${HZ_VERSION}.tar.gz'
```

File: .github/workflows/publish-packages.yml [hunk 8]
```diff
@@ -25,11 +25,17 @@ jobs:
       ARTIFACTORY_SECRET: ${{ secrets.ARTIFACTORY_SECRET }}
       DEVOPS_PRIVATE_KEY: ${{ secrets.DEVOPS_PRIVATE_KEY }}
       BINTRAY_PASSPHRASE: ${{ secrets.BINTRAY_PASSPHRASE }}
+    defaults:
+      run:
+        working-directory: ./hazelcast-command-line
```
Comment (ldziedziul) [inline .github/workflows/publish-packages.yml:30]: nice 👍

**Review by ldziedziul — COMMENTED** [review]
Looks good, lest just one comment

**Review by ldziedziul — APPROVED** [review]
awesome work!

**Comment by frant-hartm** [comment]
All the issues have been addressed.

I have opened another PR #113 because we can't access secrets from a fork. That PR is from a branch in this repository and can run the automation properly.

For this reason, all PRs which need to run the automation will have to be created from this repository's branches.

**Pull request closed without merging**